src/behavior-considered-undefined.md

```diff
@@ -51,7 +51,17 @@ code.
 * Invoking undefined behavior via compiler intrinsics.
 * Executing code compiled with platform features that the current platform
   does not support (see [`target_feature`]), *except* if the platform explicitly documents this to be safe.
-* Calling a function with the wrong call ABI or unwinding from a function with the wrong unwind ABI.
+* Calling a function with the wrong call ABI or unwinding from a function with the wrong unwind ABI
+* Calling a foreign (e.g. C++) function that unwinds (`throw`s) via a function
+  declaration or pointer declared with a non-unwinding ABI such as `"C"`
+* Calling a Rust `extern` function that unwinds (with `extern "C-unwind"` or
+  another ABI that permits unwinding) from a runtime that does not support
+  unwinding, such as code compiled with GCC or Clang using `-fno-exceptions`
+* Catching a Rust `panic` in non Rust code (for instance `catch (...)` in C++)
+* Catching a non-Rust unwind (such as a C++ exception) with `catch_unwind`
```

Comment:
Does this have to be UB? It would be safer to wrap the exception from the non-Rust unwind when it reaches Rust, so that `catch_unwind` just works.

After all, the point of `catch_unwind` is to catch *all* possible unwinding, and the point of using unwinding ABIs is to make unwinding from foreign code into Rust safe. So if the combination of these features is not safe, then it arguably creates a *new* footgun.

Reply:
Yes, for now, it does need to be undefined, and it may be that we need to introduce a new `catch` mechanism that is more universal (this bothers me, too). There are implementation complexities I can't recall at the moment. I can take a look later in our Zulip channel and/or project-group notes to recall the specifics.

Reply:
I agree that `"C-unwind"` and `catch_unwind` sound like they should work together, and in conjunction they're a footgun. 😔

Reply:
What is it about how `catch_unwind` is implemented that makes this undefined? Note that `catch_unwind` doesn't allow you to get the "thrown" object associated with the unwind, and so naively I would have thought that it doesn't matter what language the unwind originates in; we're not trying to interpret an object thrown from another language.

(I don't know whether the "thrown" object is used to implement the payload from a panic. It doesn't need to be; you could use a thread-local to set the payload. Although I guess that might be an incompatible ABI change.)

Reply:
Here's the existing Zulip thread: https://rust-lang.zulipchat.com/#narrow/stream/210922-project-ffi-unwind/topic/Allowing.20catch_unwind.20for.20foreign.20exceptions/near/284872034

Complications, based on reviewing that thread:

* Catching a C++ exception is, for some reason I don't quite understand, UB on Itanium unless a specific C++ standard library function is called.
* `catch_unwind` does actually allow you to get the "thrown" object (it's the error variant of the result), but presumably we could do something to prevent users from actually downcasting from `&dyn Any` to get the foreign exception object. (And yes, even with a pure Rust panic, you can throw objects other than `&str`, using `std::panic::panic_any`.)

Reply:
I am going to resolve this; feel free to reopen if you have another suggestion for how the documentation should be changed.